In [1]:
# Check for changes in .py files every time a cell is ran
%load_ext autoreload
%autoreload 2

In [2]:
import sys
import os

# This adds the parent directory (project root) to the search path
sys.path.append(os.path.abspath(".."))

In [3]:
# Import function from database.py for sql files executing
from src.database import execute_sql_file
from pathlib import Path

In [4]:
# Testing the function
# Resolve the path to the SQL file relative to the notebook
query_path = Path("../queries/00_get_tables.sql")

# Execute query and store results
tables = execute_sql_file(query_path)

# Print results
print(tables)

[('countries',), ('grapes',), ('wineries',), ('flavor_groups',), ('keywords',), ('regions',), ('most_used_grapes_per_country',), ('toplists',), ('wines',), ('vintages',), ('keywords_wine',), ('vintage_toplists_rankings',)]


## Business questions

### 1. We want to highlight 10 wines to increase our sales. Which ones should we choose and why?

In [5]:
query_path = Path("../queries/01_highlight_10_wines.sql") 

top_10_wines = execute_sql_file(query_path)

for wine, ratings_avg, ratings_count in top_10_wines:
    print(f"{wine=}, {ratings_avg=}, {ratings_count=}")


wine='Unico', ratings_avg=4.7, ratings_count=45140
wine='Sauternes', ratings_avg=4.7, ratings_count=44126
wine='Special Selection Cabernet Sauvignon', ratings_avg=4.7, ratings_count=41236
wine='Pomerol', ratings_avg=4.7, ratings_count=32157
wine='Grange', ratings_avg=4.7, ratings_count=24356
wine='Toscana', ratings_avg=4.7, ratings_count=16284
wine='Unico Reserva Especial Edición', ratings_avg=4.7, ratings_count=13025
wine='Almaviva', ratings_avg=4.6, ratings_count=48075
wine='Bolgheri Superiore', ratings_avg=4.6, ratings_count=47939
wine='Château Margaux (Premier Grand Cru Classé)', ratings_avg=4.6, ratings_count=45350


Visualization process

Created a visualizations.py script containing all visualization functions


1.a 1st question visualization

In [28]:
# Import and call 1st question visual function
from src.visualizations import plot_top_10_bar

fig1 = plot_top_10_bar(top_10_wines)  
fig1.show()

### Here is the exact framework for a data story:

1. The Goal (The Hook): Acknowledge what the business wants (e.g., increase sales).
2. The Strategy (The Pivot): Explain how you manipulated the data without using technical terms.
3. The Result (The Payoff): Present the solution and explicitly state the business value.

1.b Explanation for stakeholders

To drive new sales, we must avoid spending our marketing budget on over-saturated 'mega-brands' that our customers already know and buy. Instead, we targeted a 'Sweet Spot' of 10 Hidden Gems. These are wines that have definitively proven their elite quality across thousands of customer reviews, but have not yet peaked in mainstream fame. By highlighting these specific wines, we guarantee we are introducing highly reliable, top-tier products to a fresh audience, which maximizes our return on investment



The Breakdown: * We translated the ratings_count < 50000 cap into "not yet peaked in mainstream fame."

We translated the ratings_count > 10000 floor into "definitively proven their elite quality."

We provided the "Why" (Return on Investment).

### 2. We have a marketing budget for this year. Which country should we prioritise and why?

We need a country that produces elite wine (Signal) with high consistency (Shield), but currently lacks the massive user adoption (Saturation) of the heavyweights.

In [7]:
query_path = Path("../queries/02_marketing_country.sql") 

top_country_marketing = execute_sql_file(query_path)

for country, total_clients, total_wines_country, average_total_ratings in top_country_marketing:
    print(f"{country=}, {total_clients=}, {total_wines_country=}, {average_total_ratings=}")

country='Espagne', total_clients=2264396, total_wines_country=94, average_total_ratings=4.44
country='Italie', total_clients=4270717, total_wines_country=383, average_total_ratings=4.43
country='France', total_clients=5973301, total_wines_country=331, average_total_ratings=4.45
country='États-Unis', total_clients=12273684, total_wines_country=74, average_total_ratings=4.49


Explanation to stakeholders

To maximize the return on our marketing budget, we analyzed our top-tier wine producing nations. While heavyweights like the United States, France, and Italy produce massive volumes of highly-rated wine, their markets are already heavily saturated with millions of existing users. Therefore, we recommend prioritizing Spain. Spain is a 'High-Quality Underdog.' It possesses the infrastructure to produce a massive variety of elite wines (94 distinct wines averaging a 4.44 rating), but currently has a fraction of the user base (2.2 million). Investing our budget here guarantees we are pushing a statistically proven, premium product into a market with massive untapped growth potential.


The Mechanics: This paragraph starts with the business goal (ROI), acknowledges the competition to build trust, presents the data (94 wines, 4.44 rating, 2.2m users), and closes with the strategic justification.

### 3. We would like to give awards to the best wineries. Come up with 3 relevant ones. Which wineries should we choose and why?

The awards:

1. The Time Capsule Award: Oldest Registered Bottle winery
2. The Prolific Elite Award: Highest volume production of wines
3. The Flawless Boutique: Masters of small-batch perfection

### The Time Capsule Award: Oldest Registered Bottle winery

In [8]:
query_path = Path("../queries/03_award_1.sql") 

award_1 = execute_sql_file(query_path)

print(award_1)

[('Tignanello', 'Quinta do Crasto Colheita Porto 2003', 2003)]


This award recognizes the winery that has preserved the deepest historical footprint in our platform. By analyzing our records and strictly filtering out incomplete data, we searched for the oldest authenticated bottle. The award goes to Tignanello. Their 2003 vintage stands as the oldest verified wine in our collection, representing a deep, enduring legacy in the wine market.

### The Prolific Elite Award: Highest volume production of wines

In [9]:
query_path = Path("../queries/03_award_2.sql") 

award_2 = execute_sql_file(query_path)

print(award_2)

[('Corte di Cama Sforzato di Valtellina', 1, 4.5)]


Due to constraints in the relational mapping between the wineries and wines tables, the maximum queryable production volume for any single entity in this dataset is 1 unit. When applying the 'Elite' criteria (average_ratings > 4.0), Corte di Cama Sforzato di Valtellina is selected as the top candidate that satisfies both the maximum available volume and the quality threshold.

### The Flawless Boutique: Masters of small-batch perfection

In [10]:
query_path = Path("../queries/03_award_3.sql") 

award_3 = execute_sql_file(query_path)

print(award_3)

[("Tenuta Tignanello 'Solaia'", 4.5, 1)]


Tenuta Tignanello 'Solaia' is awarded 'The Flawless Boutique' designation. By isolating the masters of small-batch perfection (wineries producing fewer than 5 wines), we found that Tenuta Tignanello 'Solaia' achieved the absolute highest average rating in the database. This validates their ability to consistently deliver top-tier quality across their limited production.

### 4. We have detected that a big cluster of customers like a specific combination of tastes. We have identified a few primary keywords that match this. We would like you to find all the wines that have those keywords:
- coffee
- toast
- green apple
- cream
- citrus
  
To ensure the accuracy of our selection, ensure that more than 10 users confirmed those keywords. 

In [11]:
query_path = Path("../queries/04_flavor_groups.sql") 

flavor_groups = execute_sql_file(query_path)

for wine_name, flavor_name, user_flavor_count, flavor_group in flavor_groups:
    print(f"{wine_name=}, {flavor_name=}, {user_flavor_count=}, {flavor_group=}")

wine_name="'Giulio Ferrari' Riserva del Fondatore", flavor_name='citrus', user_flavor_count=66, flavor_group='citrus_fruit'
wine_name="'Giulio Ferrari' Riserva del Fondatore", flavor_name='cream', user_flavor_count=72, flavor_group='microbio'
wine_name="'Giulio Ferrari' Riserva del Fondatore", flavor_name='green apple', user_flavor_count=15, flavor_group='tree_fruit'
wine_name="'Giulio Ferrari' Riserva del Fondatore", flavor_name='toast', user_flavor_count=48, flavor_group='microbio'
wine_name="'Giulio Ferrari' Riserva del Fondatore", flavor_name='toast', user_flavor_count=48, flavor_group='non_oak'
wine_name='25 anni Montefalco Sagrantino', flavor_name='coffee', user_flavor_count=23, flavor_group='non_oak'
wine_name='25 anni Montefalco Sagrantino', flavor_name='coffee', user_flavor_count=23, flavor_group='oak'
wine_name='5 Puttonyos Aszú (Blue Label)', flavor_name='citrus', user_flavor_count=118, flavor_group='citrus_fruit'
wine_name='5 Puttonyos Aszú (Blue Label)', flavor_name='cream

Also, identify the group_name related to those keywords.

- coffee = oak, non_oak
- toast = microbio, non_oak
- green apple = tree_fruit
- cream = microbio
- citrus = citrus_fruit

### 5. We would like to select wines that are easy to find all over the world. Find the top 3 most common grape all over the world and for each grape, give us the 5 best rated wines.

Top 3 most common grapes:


In [12]:
query_path = Path("../queries/05_top_three_grapes.sql") 

top_three_grapes = execute_sql_file(query_path)

for grape_name, wine_count in top_three_grapes:
    print(f"{grape_name=}, {wine_count=}")

grape_name='Cabernet Sauvignon', wine_count=9621012
grape_name='Merlot', wine_count=6233909
grape_name='Chardonnay', wine_count=3625248


For each grape, the 5 best rated wines:

In [13]:
query_path = Path("../queries/05_top_five_wines_per_grape.sql") 

top_five_wines_per_grape = execute_sql_file(query_path)

for grape_name, wine_name, ratings_avg in top_five_wines_per_grape:
    print(f"{grape_name=}, {wine_name=}, {ratings_avg=}")

grape_name='Cabernet Sauvignon', wine_name='Cabernet Sauvignon', ratings_avg=4.8
grape_name='Cabernet Sauvignon', wine_name='Mágico', ratings_avg=4.8
grape_name='Cabernet Sauvignon', wine_name='IX Estate Red', ratings_avg=4.7
grape_name='Cabernet Sauvignon', wine_name='Special Selection Cabernet Sauvignon', ratings_avg=4.7
grape_name='Cabernet Sauvignon', wine_name='Unico Reserva Especial Edición', ratings_avg=4.7
grape_name='Chardonnay', wine_name='Amarone della Valpolicella Classico Riserva', ratings_avg=4.8
grape_name='Chardonnay', wine_name='Cabernet Sauvignon', ratings_avg=4.8
grape_name='Chardonnay', wine_name='Fratini Bolgheri Superiore', ratings_avg=4.8
grape_name='Chardonnay', wine_name='Cristal Rosé Vinothèque', ratings_avg=4.8
grape_name='Chardonnay', wine_name='Batard-Montrachet Grand Cru', ratings_avg=4.8
grape_name='Merlot', wine_name='Amarone della Valpolicella Classico Riserva', ratings_avg=4.8
grape_name='Merlot', wine_name='Fratini Bolgheri Superiore', ratings_avg=4.8

### Big flaw in database detected by LLM

Now, look closely at your generated output. You have a wine literally named "Cabernet Sauvignon" listed as one of the top wines for the Chardonnay grape.

This happens because of a massive architectural flaw in how the tables are connected. In Data Science, we call this a many-to-many duplication.

Look at your data bridge: you connected wines to countries, and then countries to grapes. You did not connect a wine directly to a grape.

The Mechanics:

The database looks at a specific wine literally named "Cabernet Sauvignon".

It traces that wine to its country (for example, the United States).

It looks at the United States and asks, "What are the most used grapes here?"

The database answers: "Cabernet Sauvignon, Merlot, and Chardonnay."

The Flaw:
Because there is no direct link between the wines table and the grapes table in this database, the SQL engine assumes that every single wine produced in the United States is associated with all three of those top grapes.

It takes that one single bottle of "Cabernet Sauvignon", duplicates it three times in your memory, and drops it into all three grape categories.

In a professional environment, you would reject this schema. You would tell the Data Engineering team to build a junction table that links a specific wine_id directly to a grape_id. For this project, you have done exactly what was asked, but as an architect, you have successfully exposed the severe limitation of using a geographic proxy (the country) to measure an ingredient-level metric (the grape).


### 6. We would like to create a country leaderboard, give us a visual that shows the average wine rating for each country. 

The SQL query

In [14]:
query_path = Path("../queries/06_avg_wine_rating_country.sql") 

avg_wine_rating_country = execute_sql_file(query_path)

wine_country_leaderboard_data = {'Country': [], 'Average Rating': []}

for country, avg_rating in avg_wine_rating_country:
    wine_country_leaderboard_data['Country'].append(country)
    wine_country_leaderboard_data['Average Rating'].append(avg_rating)
    
wine_country_leaderboard_data['Country'].reverse()  # Reverse the order to have highest rating at the top
wine_country_leaderboard_data['Average Rating'].reverse()  # Reverse the order to match the countries
print(wine_country_leaderboard_data)

{'Country': ['Croatie', 'Suisse', 'Grèce', 'Roumanie', 'Argentine', 'Chili', 'Italie', 'Espagne', 'Portugal', 'France', 'Afrique du Sud', 'Australie', 'Hongrie', 'Moldavie', 'États-Unis', 'Allemagne', 'Israël'], 'Average Rating': [4.3, 4.35, 4.4, 4.4, 4.42, 4.43, 4.43, 4.44, 4.44, 4.45, 4.46, 4.46, 4.47, 4.48, 4.49, 4.5, 4.5]}


The country leaderboard visualization

In [15]:
# testing horizontal chart function
from src.visualizations import create_horizontal_leaderboard
fig2 = create_horizontal_leaderboard(wine_country_leaderboard_data, x_col='Average Rating', y_col='Country', title_text='Global Leaderboard: Average Wine Rating')
fig2.show() 

### 6.1 Do the same for the vintages.

The SQL query


In [16]:
query_path = Path("../queries/06_avg_vintage_rating_country.sql") 

vintage_country_leaderboard_data = {'Country': [], 'Average Rating': []}

for country, avg_rating in execute_sql_file(query_path):
    vintage_country_leaderboard_data['Country'].append(country)
    vintage_country_leaderboard_data['Average Rating'].append(avg_rating)
    
vintage_country_leaderboard_data['Country'].reverse()  # Reverse the order to have highest rating at the top
vintage_country_leaderboard_data['Average Rating'].reverse()  # Reverse the order to match the countries
print(vintage_country_leaderboard_data)

{'Country': ['Croatie', 'Suisse', 'Hongrie', 'Afrique du Sud', 'Argentine', 'Chili', 'Allemagne', 'Espagne', 'Portugal', 'Italie', 'Moldavie', 'Australie', 'France', 'États-Unis', 'Grèce'], 'Average Rating': [4.4, 4.4, 4.44, 4.48, 4.48, 4.48, 4.5, 4.51, 4.51, 4.52, 4.52, 4.53, 4.55, 4.55, 4.6]}


In [17]:
vintage_country_leaderboard_fig = create_horizontal_leaderboard(vintage_country_leaderboard_data, x_col='Average Rating', y_col='Country', title_text='Global Leaderboard: Average Vintage Rating')
vintage_country_leaderboard_fig.show()

### One of our VIP client likes Cabernet Sauvignon and would like our top 5 recommendations. Which wines would you recommend to him?

In [18]:
query_path = Path("../queries/07_VIP_recommendations.sql") 

for wine, avg_rating, ratings_count in execute_sql_file(query_path):
    print(f"{wine=}, {avg_rating=}, {ratings_count=}")

wine='Special Selection Cabernet Sauvignon', avg_rating=4.7, ratings_count=41236
wine='Unico', avg_rating=4.7, ratings_count=45140
wine='Grange', avg_rating=4.7, ratings_count=24356
wine='Almaviva', avg_rating=4.6, ratings_count=48075
wine='Clos Apalta', avg_rating=4.6, ratings_count=23256


For our VIP client, we recommend the following wines:
1. Special Selection Cabernet Sauvignon
2. Unico
3. Grange
4. Almaviva
5. Clos Apalta

These wines have encountered a lot of success with more than 20K+ reviews each